src/main/resources/META-INF/services/org.apache.lucene.codecs.Codec

```diff
@@ -1,3 +1,5 @@
-org.opensearch.index.codec.customcodecs.ZstdCodec
-org.opensearch.index.codec.customcodecs.ZstdNoDictCodec
-org.opensearch.index.codec.customcodecs.ZstdDeprecatedCodec
+org.opensearch.index.codec.customcodecs.backward_codecs.Zstd95Codec
+org.opensearch.index.codec.customcodecs.backward_codecs.ZstdNoDict95Codec
+org.opensearch.index.codec.customcodecs.backward_codecs.Zstd95DeprecatedCodec
+org.opensearch.index.codec.customcodecs.Zstd99Codec
+org.opensearch.index.codec.customcodecs.ZstdNoDict99Codec
```

Comment:
I don't think we can safely change these. 

Codecs are loaded [via reflection](https://github.com/apache/lucene/blob/f16007c3eca7cbf89bf6c2f88907707cf30c0058/lucene/core/src/java/org/apache/lucene/index/SegmentInfos.java#L500). The segment infos holds the full class name.

I think you should be able to test this manually by writing a little tool in this package that writes an index. (I would just use `IndexWriter` directly with an `IndexWriterConfig` that uses your custom codec. [Here's an example](https://msfroh.github.io/lucene-university/docs/DirectoryFileContents.html).) 

1. Write an index without this commit. That will simulate the "2.11 behavior".
2. Apply this commit. Open an `IndexReader` over the 2.11 index.

I believe that you **must* keep the old fully-qualified class name to maintain backward compatibility (which unfortunately doesn't mention the version). Going forward, you can+should include version numbers.

Reply:
> Codecs are loaded [via reflection](https://github.com/apache/lucene/blob/f16007c3eca7cbf89bf6c2f88907707cf30c0058/lucene/core/src/java/org/apache/lucene/index/SegmentInfos.java#L500). The segment infos holds the full class name.

Hm ... shouldn't it be code name (not class name)? Lucene moves codec under `backward_codecs` package every time, how does it work there?

Reply:
My understanding as well is that the provided name within the default constructor for codec class implementation is being used (based on how NamedSPILoader is being used in Lucene).

https://github.com/apache/lucene/blob/3a205feb27de8450c3bac7242f3c976ce5d21436/lucene/core/src/java/org/apache/lucene/codecs/Codec.java#L66-L68

Reply:
I tested this by restoring the data indexed with zstd codecs in OSv2.11 to OSv3.0. Was able to ingest, and search after moving the codecs to the ```backward_codecs``` folder.
Here is part of the output when I hit ```GET <index>/_segments``` locally, was able to see both old and new segments (before and after the lucene upgrade).

```
                        "segments": {
                            "_f": {
                                "generation": 15,
                                "num_docs": 1767,
                                "deleted_docs": 0,
                                "size_in_bytes": 53849,
                                "memory_in_bytes": 0,
                                "committed": false,
                                "search": true,
                                "version": "9.7.0",
                                "compound": false,
                                "attributes": {
                                    "Lucene95CustomStoredFieldsFormat.mode": "ZSTD_NO_DICT"
                                }
                            },
                            "_v": {
                                "generation": 31,
                                "num_docs": 432,
                                "deleted_docs": 0,
                                "size_in_bytes": 14885,
                                "memory_in_bytes": 0,
                                "committed": false,
                                "search": true,
                                "version": "9.9.1",
                                "compound": true,
                                "attributes": {
                                    "Lucene99CustomStoredFieldsFormat.mode": "ZSTD_NO_DICT"
                                }
                            },
```

Reply:
Ahh... got it. Thanks @mgodwan for the link to the codec comment -- I had mistakenly assumed the FQCN got written to SegmentInfos.

In that case, we should be good to go once we have the BWC test. 👍